Initial Storybook setup

jojopirker opened the pull request:
These changes enable to use `npm run storybook` to start **storybook** as mentioned in #102. 
There is an example story for the header.
Maybe should be moved to the components folder when doing #99.

What is **not finished**: 
- DataMocking
- fokus only on `components/` rather than `pages/`

What is **finished**:
- TailwindCSS and ChakraUI integration

fozziethebeat commented on website/src/stories/Header.stories.jsx:
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import { SessionContext } from 'next-auth/react';
+
+
```
style nit, can you delete this newline?

fozziethebeat commented on website/src/stories/Header.stories.jsx:
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import { SessionContext } from 'next-auth/react';
```
Style nit, can you order these alphabetically?

fozziethebeat commented on website/src/stories/Header.stories.jsx:
```diff
@@ -0,0 +1,26 @@
+import { SessionContext } from 'next-auth/react';
+
+
+import { Header } from '../components/Header';
```
Is it possible to import from `src/components/Header`? If not, that's okay

jojopirker replied:
moved them into the same folder, think thats inline with #99

fozziethebeat commented:
Great first step! I have a few small requested changes then I'll merge this in.

jojopirker commented:
Deleted the empty lines :)

I also moved the files so its more inline with future structure.

AbdBarho commented:
would make sense the have the stories in another folder? i.e. not between all the components? maybe in `src/stories/...`

I am suggesting this because stories are more of "write and forget", otherwise the files would be in the way all the time.

@jojopirker @fozziethebeat  what do you think?

fozziethebeat commented:
> would make sense the have the stories in another folder? i.e. not between all the components? maybe in `src/stories/...`
> 
> I am suggesting this because stories are more of "write and forget", otherwise the files would be in the way all the time.
> 
> @jojopirker @fozziethebeat what do you think?

This is definitely a matter of preference and probably past experiences.  

I like how RedwoodJS put the test and stories in the same directory as the component so that when you make changes to the component you're reminded to check the story or test and don't have to go find the corresponding file in some other parallel directory structure.  Let's try this for now and then re-evaluate.

fozziethebeat commented on website/src/pages/index.tsx:
```diff
@@ -7,7 +7,7 @@ import { CallToAction } from "src/components/CallToAction";
 import { Footer } from "src/components/Footer";
 import { Header } from "src/components/Header";
-import { Hero } from "src/components/Hero";
+import { Hero } from "sr../components/Header/Header
```
This looks wrong, I think you meant to change line 9

fozziethebeat commented on website/src/components/Header/Header.stories.jsx:
```diff
@@ -0,0 +1,25 @@
+  title: 'Example/Header',
+  component: Header,
+  parameters: {
+    // More on Story layout: https://storybook.js.org/docs/react/configure/story-layout
```
we can probably delete this in the code.  If people get confused we can add explanations in a README.

fozziethebeat commented on website/src/components/Header/Header.tsx:
```diff
@@ -6,9 +6,9 @@ import Link from "next/link";
-import { Avatar } from "./Avatar";
-import { Container } from "./Container";
-import { NavLinks } from "./NavLinks";
+import { Avatar } from "../Avatar/Avatar";
```
Can these be absolute imports (`src/components/Avatar/Avatar`)?

fozziethebeat reviewed (approved):
Looks good after a few tiny changes and running prettier.

fozziethebeat reviewed (changes requested):
Holding off on the approval until the merge fixes are in.  I'd like to make sure those are all addressed and no new issues get introduced.

jojopirker commented:
Should include all the fixes now.

AbdBarho commented on website/src/pages/create/summarize_story.tsx:
```diff
@@ -5,7 +5,7 @@ import useSWRImmutable from "swr/immutable";
 
 import { Footer } from "src/components/Footer";
-import { Header } from "src/components/Header";
+import { Header } from "@/components/Header/Header";
```
can we replace `@` with `src`?
I know vscode automatically added it....

jojopirker replied:
I removed the line entirely while merging..

AbdBarho commented on website/src/pages/evaluate/rate_summary.tsx:
```diff
@@ -7,7 +7,7 @@ import useSWRImmutable from "swr/immutable";
 
 import { Footer } from "src/components/Footer";
-import { Header } from "src/components/Header";
+import { Header } from "@/components/Header/Header";
```
same here

jojopirker replied:
also gone :D

AbdBarho reviewed (commented):
Thanks!
Could you please also resolve the merge conflicts so we can merge this?

AbdBarho commented on website/.storybook/main.js:
```diff
@@ -0,0 +1,28 @@
+  },
+  staticDirs: ["../public"],
+  typescript: { reactDocgen: false },
+  webpackFinal: async (config, { configType }) => {
```
why is this necessary?

jojopirker replied:
` typescript: { reactDocgen: false },` was also necessary for absolute imports... i just cant find the source where i found the solution

jojopirker replied:
found [it](https://github.com/storybookjs/storybook/issues/15336#issuecomment-888528747), had an error when trying to use the absolute imports, that was a fix for it..

AbdBarho replied:
https://github.com/storybookjs/storybook/issues/15336#issuecomment-888528747

AbdBarho replied:
okay, so the reason is absolute imports, maybe you could document that.

In the link above some users reported added webpack v5 to dev dependencies and everything just worked, however, it pains me that we have to do this.

The offical package has not been updated in a while, which is a bad sign: https://www.npmjs.com/package/@storybook/react-docgen-typescript-plugin


We lose react-docgen, but I think that is ok.

For the webpack final, can you please make it explicit that it is only resolving aliases, and not replacing modules.

```js

    config.resolve.alias = {
      ...config.resolve.alias,
      src: path.resolve(__dirname, "../src"),
    };
    return config;
```

AbdBarho commented on website/.storybook/main.js:
```diff
@@ -0,0 +1,28 @@
+    return config;
+  },
+  features: {
+    emotionAlias: false,
```
why is this necessary?

jojopirker replied:
necessary for for chakra to work properly, see [here](https://chakra-ui.com/getting-started/with-storybook)

AbdBarho commented on website/.storybook/preview.js:
```diff
@@ -0,0 +1,20 @@
+import "!style-loader!css-loader!postcss-loader!tailwindcss/tailwind.css";
```
this should be replaced in the future with a theme provider there includes chakra ui & tailwind stuff.

AbdBarho commented on website/.storybook/preview.js:
```diff
@@ -0,0 +1,20 @@
+  },
+};
+
+import * as NextImage from "next/image";
+
+const OriginalNextImage = NextImage.default;
+
+Object.defineProperty(NextImage, "default", {
+  configurable: true,
+  value: (props) => <OriginalNextImage {...props} unoptimized />,
+});
```
oof...

what's the problem here exactly?

jojopirker replied:
little hacky solution because next handles images in a way, that simply adding the public folder to storybook would not work... i used the solution from [here](https://dev.to/jonasmerlin/how-to-use-the-next-js-image-component-in-storybook-1415)

AbdBarho replied:
Damn, very nasty, could you  please document this in the code? so that future developers can understand why.

AbdBarho reviewed (commented):
while installing  locally, I get the following in the console:
```sh
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @mdx-js/react@1.6.22
npm WARN Found: react@18.2.0
npm WARN node_modules/react
npm WARN   react@"18.2.0" from the root project
npm WARN   128 more (@chakra-ui/accordion, @chakra-ui/alert, ...)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peer react@"^16.13.1 || ^17.0.0" from @mdx-js/react@1.6.22
npm WARN node_modules/@mdx-js/react
npm WARN   @mdx-js/react@"^1.6.22" from @storybook/addon-docs@6.5.15
npm WARN   node_modules/@storybook/addon-docs
npm WARN
npm WARN Conflicting peer dependency: react@17.0.2
npm WARN node_modules/react
npm WARN   peer react@"^16.13.1 || ^17.0.0" from @mdx-js/react@1.6.22
npm WARN   node_modules/@mdx-js/react
npm WARN     @mdx-js/react@"^1.6.22" from @storybook/addon-docs@6.5.15
npm WARN     node_modules/@storybook/addon-docs
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: react-element-to-jsx-string@14.3.4
npm WARN Found: react@18.2.0
npm WARN node_modules/react
npm WARN   react@"18.2.0" from the root project
npm WARN   128 more (@chakra-ui/accordion, @chakra-ui/alert, ...)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peer react@"^0.14.8 || ^15.0.1 || ^16.0.0 || ^17.0.1" from react-element-to-jsx-string@14.3.4
npm WARN node_modules/react-element-to-jsx-string
npm WARN   react-element-to-jsx-string@"^14.3.4" from @storybook/react@6.5.15
npm WARN   node_modules/@storybook/react
npm WARN 
npm WARN Conflicting peer dependency: react@17.0.2
npm WARN node_modules/react
npm WARN   peer react@"^0.14.8 || ^15.0.1 || ^16.0.0 || ^17.0.1" from react-element-to-jsx-string@14.3.4
npm WARN   node_modules/react-element-to-jsx-string
npm WARN     react-element-to-jsx-string@"^14.3.4" from @storybook/react@6.5.15
npm WARN     node_modules/@storybook/react
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: react-element-to-jsx-string@14.3.4
npm WARN Found: react-dom@18.2.0
npm WARN node_modules/react-dom
npm WARN   react-dom@"18.2.0" from the root project
npm WARN   40 more (@chakra-ui/modal, @chakra-ui/portal, ...)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peer react-dom@"^0.14.8 || ^15.0.1 || ^16.0.0 || ^17.0.1" from react-element-to-jsx-string@14.3.4
npm WARN node_modules/react-element-to-jsx-string
npm WARN   react-element-to-jsx-string@"^14.3.4" from @storybook/react@6.5.15
npm WARN   node_modules/@storybook/react
npm WARN
npm WARN Conflicting peer dependency: react-dom@17.0.2
npm WARN node_modules/react-dom
npm WARN   peer react-dom@"^0.14.8 || ^15.0.1 || ^16.0.0 || ^17.0.1" from react-element-to-jsx-string@14.3.4
npm WARN   node_modules/react-element-to-jsx-string
npm WARN     react-element-to-jsx-string@"^14.3.4" from @storybook/react@6.5.15
npm WARN     node_modules/@storybook/react
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: react-inspector@5.1.1
npm WARN Found: react@18.2.0
npm WARN node_modules/react
npm WARN   react@"18.2.0" from the root project
npm WARN   128 more (@chakra-ui/accordion, @chakra-ui/alert, ...)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peer react@"^16.8.4 || ^17.0.0" from react-inspector@5.1.1
npm WARN node_modules/react-inspector
npm WARN   react-inspector@"^5.1.0" from @storybook/addon-actions@6.5.15
npm WARN   node_modules/@storybook/addon-actions
npm WARN
npm WARN Conflicting peer dependency: react@17.0.2
npm WARN node_modules/react
npm WARN   peer react@"^16.8.4 || ^17.0.0" from react-inspector@5.1.1
npm WARN   node_modules/react-inspector
npm WARN     react-inspector@"^5.1.0" from @storybook/addon-actions@6.5.15
npm WARN     node_modules/@storybook/addon-actions
npm WARN deprecated chokidar@2.1.8: Chokidar 2 does not receive security updates since 2019. Upgrade to chokidar 3 with 15x fewer dependencies
```
Is this expected? storybook already has react 18 support, not sure why it is complaining  here.

fozziethebeat reviewed (approved):
This looks pretty good.  Please make sure the merge with header.tsx works smoothly and nothing gets dropped.

jojopirker commented:
merged the previous changes... i also moved all the files of the header into `/components/header` so they are all in one place, and added and index file in the folder for better import

AbdBarho commented:
> merged the previous changes... i also moved all the files of the header into `/components/header` so they are all in one place, and added and index file in the folder for better import

Thanks! I see a `docker-compose 2.yaml` file, this was not intended right?

jojopirker commented:
yup... i still need to learn how to properly merge with git^^
